From 51a9e8e2268fa9b34424ca814f08cbaca8f2ea37 Mon Sep 17 00:00:00 2001
From: Paul Naoumenko <paulx.naoumenko@intel.com>
Date: Tue, 15 Oct 2013 15:46:47 +0200
Subject: [PORT FROM MAIN] wifi: Don't keep hidden AP in the list after forget it

BZ: 140281

This patch removes hidden BSS that are no more connected.
Before this patch, such BSS was remaining in the list during
1 minute until aging mechanism finally removes it.

Change-Id: I96e30e609403d627b40729000f560651b9b34c25
Orig-Change-Id: I90062c52be4269439b58bbe0eac51f4e79ed417f
Category: device enablement
Domain: CWS.WIFI-Common
Origin: upstream
Upstream-Candidate: yes
Signed-off-by: Matthieu Mauger <matthieux.mauger@intel.com>
---
 wpa_supplicant/bss.c            |    4 ++++
 wpa_supplicant/bss.h            |    2 ++
 wpa_supplicant/wpa_supplicant.c |    4 ++++
 3 files changed, 10 insertions(+), 0 deletions(-)

diff --git a/wpa_supplicant/bss.c b/wpa_supplicant/bss.c
index be1632f..fa64fd7 100644
--- a/wpa_supplicant/bss.c
+++ b/wpa_supplicant/bss.c
@@ -333,6 +333,7 @@ static struct wpa_bss * wpa_bss_add(struct wpa_supplicant *wpa_s,
 	bss = os_zalloc(sizeof(*bss) + res->ie_len + res->beacon_ie_len);
 	if (bss == NULL)
 		return NULL;
+	bss->is_hidden = 0;
 	bss->id = wpa_s->bss_next_id++;
 	bss->last_update_idx = wpa_s->bss_update_idx;
 	wpa_bss_copy_res(bss, res, fetch_time);
@@ -736,6 +737,9 @@ void wpa_bss_update_end(struct wpa_supplicant *wpa_s, struct scan_info *info,
 		    wpa_s->conf->bss_expiration_scan_count) {
 			wpa_bss_remove(wpa_s, bss, "no match in scan");
 		}
+		if (bss->is_hidden) {
+			wpa_bss_remove(wpa_s, bss, "not connected hidden BSS");
+		}
 	}
 
 	wpa_printf(MSG_DEBUG, "BSS: last_scan_res_used=%u/%u "
diff --git a/wpa_supplicant/bss.h b/wpa_supplicant/bss.h
index bae01d2..debb18b 100644
--- a/wpa_supplicant/bss.h
+++ b/wpa_supplicant/bss.h
@@ -81,6 +81,8 @@ struct wpa_bss {
 	int noise;
 	/** Signal level */
 	int level;
+	/** SSID is hidden */
+	int is_hidden;
 	/** Timestamp of last Beacon/Probe Response frame */
 	u64 tsf;
 	/** Time of the last update (i.e., Beacon or Probe Response RX) */
diff --git a/wpa_supplicant/wpa_supplicant.c b/wpa_supplicant/wpa_supplicant.c
index 18b78f7..7ccf273 100644
--- a/wpa_supplicant/wpa_supplicant.c
+++ b/wpa_supplicant/wpa_supplicant.c
@@ -1756,6 +1756,10 @@ void wpa_supplicant_deauthenticate(struct wpa_supplicant *wpa_s,
 	u8 *addr = NULL;
 	union wpa_event_data event;
 	int zero_addr = 0;
+	if (wpa_s->current_ssid != NULL && wpa_s->current_ssid->scan_ssid) {
+		wpa_s->current_bss->is_hidden = 1;
+	}
+
 
 	wpa_dbg(wpa_s, MSG_DEBUG, "Request to deauthenticate - bssid=" MACSTR
 		" pending_bssid=" MACSTR " reason=%d state=%s",
-- 
1.7.4.1

